state/upgrades.go

```diff
@@ -3007,3 +3007,36 @@ func ResetDefaultRelationLimitInCharmMetadata(pool *StatePool) (err error) {
 		return errors.Trace(st.db().RunTransaction(ops))
 	}))
 }
+
+func RemoveUnusedLinkLayerDeviceProviderIDs(pool *StatePool) error {
+	st := pool.SystemState()
+
+	const idType = "linklayerdevice"
+
+	lldCol, lldCloser := st.db().GetRawCollection(linkLayerDevicesC)
+	defer lldCloser()
+
+	// Gather the full qualified IDs for used link-layer device provider IDs.
+	var used []string
+	var doc struct {
+		ModelUUID  string `bson:"model-uuid"`
+		ProviderID string `bson:"providerid,omitempty"`
+	}
+	iter := lldCol.Find(nil).Iter()
+	for iter.Next(&doc) {
+		used = append(used, strings.Join([]string{doc.ModelUUID, idType, doc.ProviderID}, ":"))
+	}
+
+	pidCol, pidCloser := st.db().GetRawCollection(providerIDsC)
+	defer pidCloser()
+
+	// Delete all link-layer device provider IDs we didn't find.
+	_, err := pidCol.RemoveAll(bson.D{{
```

Comment:
Instead of passing `used` in as-is, we might want to consider splitting it into a number of chunks (e.g 100-200 entries each or something like that) and doing a delete for each one.

If the DB contains a significant number of link layer devices, passing them all in might cause the query to timeout.

Reply:
I don't expect the *provider IDs* to be all that many - that's what we're deleting here.
Accessing the devices (where we could have many) is cursor-based, so that should be OK too...

Reply:
Yes but you are iterating the linklayerdevices collection and appending an entry to `used` for each doc in the collection. So, if you have 500K items in linklayerdevices, you will pass 500K items in the query, right? Mongo might have a hard time parsing that query.

Reply:
If we anticipate this to be a problem you can try a small hack:
- Add a `tagged` field with a value `1` to each entry in providerIDs
- For each chunk, `$set tagged:0` for the ids in the chunk
- When done, delete the ones that are still tagged.

Reply:
So you can't trivially do a batch of used, because you are saying "everything that *isn't* in used", so if you shrunk that batch you would delete things that are referenced.

I just want to confirm there isn't any reason for used to be a set, or for us to sort the list so that mongo can be more efficient about its $nin query.
(linear search through 200 items for every document in another collection sounds expensive)
The only way to do a sort of batching would be to turn used into a set in memory, then query for all provider ids in the collection, and manually pull out ones that aren't in used and then batch the removals.

Reply:
Good point. I've amended the commit so that only devices *with* a provider ID are queried.

Reply:
We don't anticipate duplicate provider ID *usage*, but I've changed it to use a set.